### PR TITLE
Wrap Route from extension in Route component

### PR DESCRIPTION
**Pull request by karelhala**
### Route component wrapper

Since extensions would like to use `useParams` and other hooks from `react-router` we have to wrap the exported component from extension with `Route` and spread the `properties` in it.

Jira: [HAC-1061](https://issues.redhat.com//browse/HAC-1061)

### CODE REVIEW

File: frontend/src/Routes/DynamicRoute/DynamicRoute.tsx [hunk 1]
```diff
@@ -16,62 +16,51 @@ type DynamicRouteProps = {
...
+  const [routes, setRoutes] = React.useState<StateRoute[]>([]);
   const dynamicRoutePages = useExtensions<DynamicRoutePage>(isDynamicRoutePage);
+
   React.useEffect(() => {
-    if (location) {
-      const { properties: currRoute, pluginName } =
-        dynamicRoutePages.find(({ properties }) => {
-          if (Array.isArray(properties.path)) {
-            return properties.path.some((path) => checkPath(location.pathname, { ...properties, path }));
+    setRoutes(dynamicRoutePages.map(({ properties: { component, ...currRoute }, pluginName }) => {
```
Comment (christianvogt) [inline frontend/src/Routes/DynamicRoute/DynamicRoute.tsx:28]: use `useMemo` and remove the `routes` state.

File: frontend/src/Routes/DynamicRoute/DynamicRoute.tsx [hunk 2]
```diff
@@ -16,62 +16,51 @@ type DynamicRouteProps = {
...
+
+  console.log(routes, 'this is routes');
 
   return (
-    <section className={currClassName || ''}>
-      {Component ? (
-        <React.Suspense fallback={null}>
+    <React.Suspense fallback={null}>
+      {routes.map(({ className, Component, ...currCoute }) => <React.Fragment>
+        <Route
```
Comment (christianvogt) [inline frontend/src/Routes/DynamicRoute/DynamicRoute.tsx:57]: add a `key` based on the extension `uid`

File: frontend/src/Routes/DynamicRoute/DynamicRoute.tsx [hunk 3]
```diff
@@ -16,62 +16,51 @@ type DynamicRouteProps = {
...
+          {...currCoute}
+          render={() => <section className={className || ''}>
```
Comment (christianvogt) [inline frontend/src/Routes/DynamicRoute/DynamicRoute.tsx:59]:
```suggestion
          render={() => <section className={className}>
```

Comment (karelhala) [reply frontend/src/Routes/DynamicRoute/DynamicRoute.tsx]: This could lead to `undefined` className.

File: frontend/src/Routes/DynamicRoute/DynamicRoute.tsx (hunk 2 above)
Comment (karelhala) [reply frontend/src/Routes/DynamicRoute/DynamicRoute.tsx]: Didn't know there's `uid`! Good point, let's lmit the number of re-renders.

File: frontend/src/Routes/DynamicRoute/DynamicRoute.tsx (hunk 1 above)
Comment (karelhala) [reply frontend/src/Routes/DynamicRoute/DynamicRoute.tsx]: I tent to avoid `useMemo`, but it's a good fit for this one. What is your opinion on `useEffect` + `useRef`, it'll basically do the same thing but without the pitfall of React ignoring some memoized values as mantioned in the docs[1]

[1] https://reactjs.org/docs/hooks-reference.html#usememo

File: frontend/src/Routes/DynamicRoute/DynamicRoute.tsx (hunk 3 above)
Comment (christianvogt) [reply frontend/src/Routes/DynamicRoute/DynamicRoute.tsx]: React will just ignore undefined values set on a prop though so it should be fine

File: frontend/src/Routes/DynamicRoute/DynamicRoute.tsx (hunk 1 above)
Comment (christianvogt) [reply frontend/src/Routes/DynamicRoute/DynamicRoute.tsx]: always `useMemo` when you want the value to be computed in the render pass. `useEffect` is run _after_ render and therefore should only be used when computing a side effect or something that requires async computation. 

`useEffect` + `useState` will result in 2 render passes vs `useMemo`.
`useEffect` + `useRef` would be a special case because updating the ref doesn't cause a re-render at all.

Comment (christianvogt) [reply frontend/src/Routes/DynamicRoute/DynamicRoute.tsx]: Also, if the code can be written entirely in the render without any hook, then `useMemo` is the appropriate choice because it's only there to improve performance.

File: frontend/src/Routes/DynamicRoute/DynamicRoute.tsx (hunk 3 above)
Comment (karelhala) [reply frontend/src/Routes/DynamicRoute/DynamicRoute.tsx]: Mhh I could swear I've seen `undefined` classes in the past. But can't reproduce it on this. Will apply the change.

File: frontend/src/Routes/DynamicRoute/DynamicRoute.tsx [hunk 4]
```diff
@@ -57,21 +43,26 @@ const DynamicRoute: React.FC<DynamicRouteProps> = ({ location }) => {
...
-        <Loader />
-      )}
-    </section>
+    <React.Suspense fallback={null}>
+      {routes.map(({ className, Component, uid, ...currCoute }) => (
+        <React.Fragment key={uid}>
+          <Route
+            {...currCoute}
+            render={() => (
+              <section className={className}>
```
Comment (christianvogt) [inline frontend/src/Routes/DynamicRoute/DynamicRoute.tsx:58]: Won't having this intermediate `section` prevent pages from being able to create content that consumes the full height of the page content? In most cases it's probably ok because the content wants the page to own the scrolling. 
But in other cases the content may want to consume the full height and control their own layout. But now the content will be confined within a block element with no defined height.

File: frontend/src/Routes/DynamicRoute/DynamicRoute.tsx [hunk 5]
```diff
@@ -57,21 +43,26 @@ const DynamicRoute: React.FC<DynamicRouteProps> = ({ location }) => {
...
-          <Component />
-        </React.Suspense>
-      ) : (
-        <Loader />
-      )}
-    </section>
+    <React.Suspense fallback={null}>
+      {routes.map(({ className, Component, uid, ...currCoute }) => (
+        <React.Fragment key={uid}>
+          <Route
```
Comment (christianvogt) [inline frontend/src/Routes/DynamicRoute/DynamicRoute.tsx:55]: The fragment is unnecessary here. You can put the `key` on the `Route` instead.

File: frontend/src/Routes/DynamicRoute/DynamicRoute.tsx (hunk 4 above)
Comment (karelhala) [reply frontend/src/Routes/DynamicRoute/DynamicRoute.tsx]: Ah good point, let's use `article` instead of `section` so we don't use a block element.

File: frontend/src/Routes/DynamicRoute/DynamicRoute.tsx (hunk 5 above)
Comment (karelhala) [reply frontend/src/Routes/DynamicRoute/DynamicRoute.tsx]: Ah, forgot to add it from previous request.

File: frontend/src/Routes/DynamicRoute/DynamicRoute.tsx (hunk 4 above)
Comment (karelhala) [reply frontend/src/Routes/DynamicRoute/DynamicRoute.tsx]: Although, `article` is `block` element as well. Mhh, this is not configurable from the app side as well (since we do wrap each CSS rule with the app name). I think it's fine for now and let's keep an eye on future requests. There's already an issue opened to fix the problematic namespacing of all CSS rules and not an option to style parent components.

Comment (christianvogt) [reply frontend/src/Routes/DynamicRoute/DynamicRoute.tsx]: My point wasn't about a block or inline element but the fact that there is an intermediate node inserted that affects the ability of the child node to grow to the size of the content container through css.

It's fine for now, but be aware that this can affect page layouts that want to consume all available space.

**Comment by karelhala** [comment]
@christianvogt any objections or, is this good to go?

**Comment by christianvogt** [comment]
Sorry @karelhala, I was on PTO.
I haven't run the PR, but overall it looks good.
My comment about insertion of an intermediate node can still affect layout of pages though.

**Comment by karelhala** [comment]
@christianvogt as for the pages layout, we'll keep an eye on it. We also have a ticket to change slightly the namespacing of classes in platform-experience, so that would remove the need for additional element.

File: frontend/src/Routes/DynamicRoute/DynamicRoute.tsx [hunk 6]
```diff
@@ -15,38 +15,24 @@ type DynamicRouteProps = {
...
-          React.lazy(async () => {
+
+  const routes = React.useMemo(
+    () =>
+      dynamicRoutePages.map(({ properties: { component, ...currRoute }, pluginName, uid }) => {
+        return {
+          ...currRoute,
+          uid,
+          className: camelCase(pluginName),
+          Component: React.lazy(async () => {
```
Comment (christianvogt) [inline frontend/src/Routes/DynamicRoute/DynamicRoute.tsx:28]: @karelhala @vojtechszocs 
if the `dynamicRoutePages` changes, will the use of `React.lazy` here create a new component which causes the old one to get unmounted an re-rendered? If so, we may need to memoize this value.

Comment (vojtechszocs) [reply frontend/src/Routes/DynamicRoute/DynamicRoute.tsx:28]: @christianvogt @karelhala

> if the `dynamicRoutePages` changes, will the use of `React.lazy` here create a new component

Yes. The `Component` property above will reference a new lazy component instance.

The `component()` Promise will, however, resolve to the same "actual" React component instance as provided by dynamic plugins.

> which causes the old one to get unmounted

Yes, because the same "actual" React component is wrapped in a newly created lazy component.

> If so, we may need to memoize this value.

The "actual" React components are accessed via Promise so we can't depend on them directly since that would be an async dependency.

In OpenShift plugin SDK, the `PluginStore` guarantees that all `a.b` code references (for the given plugin) point to the same `CodeRef` instance - see `codeRefKey` and `codeRefCache` usage in [coderefs.ts](https://github.com/openshift/dynamic-plugin-sdk/blob/main/packages/lib-core/src/store/coderefs.ts).

Therefore, we could memoize lazy components based on `CodeRef` references to the "actual" components.

However, I think this would involve the use of React refs since we want to remember component `CodeRef`s across renders, instead of always deriving that information from `dynamicRoutePages` (as is the usual pattern with React hooks, e.g. get A from hookA, get B based on A from hookB etc).

**TL;DR** if this code works, I'd keep it and not worry about the extra re-renders unless we are really concerned about the performance. But ideally, yes - we should memoize all lazy component instances that defer to `CodeRef` components, assuming referential stability of `CodeRef`s. Maybe we should introduce a utility hook for this in OpenShift plugin SDK. cc @florkbr

Comment (karelhala) [reply frontend/src/Routes/DynamicRoute/DynamicRoute.tsx:28]: Good point! I'd keep it as is as well, we can track if changes to `dynamicRoutePages` causes a lot of re-renders, if it'll be the case we could optimize later. After all we want to trigger a re-render when this prop changes and react-router will do the heavy lifting of preventing the extra renders.

Comment (christianvogt) [reply frontend/src/Routes/DynamicRoute/DynamicRoute.tsx:28]: This would only affect an env with several plugins that load in asynchronously which add their own routes.

**Comment by christianvogt** [comment]
lgtm - if you want to wait on addressing the other two items that were brought up

**Comment by karelhala** [comment]
@christianvogt let's see if these two things affect the future work. We have a history of it and if we'll see these problems to come to life we'd know what to do.

**Review by rohitkrai03 — APPROVED** [review]
/lgtm
